File: sage/src/ast.cpp
#include "../include/sage.h"

using namespace ast;

Int::Int(int value) : value(value) {}
void Int::show()
{
    std::cout << value;
}

Add::Add(Node *left, Node *right) : left(left), right(right) {}
Add::~Add()
{
    delete left;
    delete right;
}
void Add::show()
{
    printf("(");
    left->show();
    printf(" + ");
    right->show();
    printf(")");
}

File: ast/ast.go
package ast

import (
	"sage/symbols"
	"sage/types"
)

type Program struct {
	Declarations []Function
	Symbols      *symbols.Table
}

type Function struct {
	Name       string
	ReturnType types.Type
	Body       Block
}

type Block struct {
	Statements []Statement
	Symbols    *symbols.Table
}

type Statement interface {
	statementNode()
}

type Return struct {
	Statement
	Value Expression
}

type Declaration struct {
	Statement
	Name  string
	Type  types.Type
	Value Expression
}

type Assignment struct {
	Statement
	Name  string
	Value Expression
}

type Expression interface {
	expressionNode()
}

type Add struct {
	Expression
	Left  Expression
	Right Expression
}

type Subtract struct {
	Expression
	Left  Expression
	Right Expression
}

type Multiply struct {
	Expression
	Left  Expression
	Right Expression
}

type Divide struct {
	Expression
	Left  Expression
	Right Expression
}

type IntegerLiteral struct {
	Expression
	Value int64
}
